### PR TITLE
Shortfin llm gpu argmax indices

### PR DESCRIPTION
**Pull request by stbaione**
- Enable using the gpu to perform argmax in the Shortfin server
- Changes the model on the `sharktank` side so that:
    - If `top-k is None`: Return the full logits
    - If `top-k == 1`: Return a subset of the logits, with the corresponding argmax indices
- Fixes a bug in the `split_argmax` implementation, adds a couple more tests for robustness/verification

Seeing marginal perf improvement when using `argmax` with `num_beams == 8`.
While not quite the jump I'd expect, it still shows promising results, as `argmax` was still pretty slow last I checked, even with the `split` implementation.

So, we're getting better results, even though the actual computation time for `argmax` is larger:

**Numpy Argmax Runs:**
```text
### Batch size: 4 ###

total requests 20
duration (s): 40.41542649269104
errors 0
mean length: 1.0

throughput (request / second): 0.49486054548049657
throughput (character / second): 0.49486054548049657
average request duration (s) 8.077639055252074
50%ile request duration (s) 8.309841752052307
75%ile request duration (s) 8.460618495941162
90%ile request duration (s) 8.737904477119447
95%ile request duration (s) 8.741231989860534
```

**Model-side Argmax Runs:**
```text
### Batch size: 4 ###

total requests 20
duration (s): 38.34156012535095
errors 0
mean length: 1.0

throughput (request / second): 0.5216271829996885
throughput (character / second): 0.5216271829996885
average request duration (s) 7.65543520450592
50%ile request duration (s) 7.543861269950867
75%ile request duration (s) 7.783713757991791
90%ile request duration (s) 8.81928973197937
95%ile request duration (s) 8.868924391269685

break: latency exceeds maximum of 5.000000s
```

### CODE REVIEW

File: sharktank/sharktank/examples/export_paged_llm_v1.py [hunk 1]
```diff
@@ -305,6 +305,14 @@ def _(model, tokens, seq_lens, seq_block_ids, cs):
             if args.logits_normalization == "log_softmax":
                 logits = ops.elementwise(torch.log, ops.softmax(logits, dim=-1))
 
+            if args.top_k is not None:
```
Comment (rsuderman) [inline sharktank/sharktank/examples/export_paged_llm_v1.py:308]: You can also do following to be more succinct
```
if args.top_k is None:
  return logits
```

File: sharktank/sharktank/examples/export_paged_llm_v1.py [hunk 2]
```diff
@@ -433,6 +441,14 @@ def _(
...
```
Comment (rsuderman) [inline sharktank/sharktank/examples/export_paged_llm_v1.py:444]: See comment above.

File: sharktank/sharktank/ops/default_impls.py [hunk 3]
```diff
@@ -49,7 +49,8 @@ def _split_argmax(input_tensor, dim, keepdim: bool = False, chunk_size: int = 12
     input_tensor = unbox_tensor(input_tensor)
     dim = dim if dim >= 0 else input_tensor.dim() + dim
 
-    tensor_split = unflatten(input_tensor, dim, (chunk_size, -1))
+    n_chunks = input_tensor.shape[dim] // chunk_size
```
Comment (rsuderman) [inline sharktank/sharktank/ops/default_impls.py:52]: Validate that the chunk size is divideable into the chunking dimension

File: shortfin/python/shortfin_apps/llm/components/batcher.py [hunk 4]
```diff
@@ -316,7 +322,13 @@ async def run(self):
                 )
 
             # Invoke VMFB. Logits are of shape [bs, bsl, d].
-            (logits,) = await fn(*args, fiber=self.fiber)
+            result = await fn(*args, fiber=self.fiber)
+
+            logits, indices = None, None
```
Comment (rsuderman) [inline shortfin/python/shortfin_apps/llm/components/batcher.py:327]:
```
logits = result[0]
if len(result) > 1:
  indices = result[1]
```

File: shortfin/python/shortfin_apps/llm/components/token_selection_strategy/greedy_token_selection_strategy.py [hunk 5]
```diff
@@ -33,9 +33,13 @@ def sample_logits(self) -> int:
         top_p = decode_config.top_p
 
         logits = np.array(exec_req.result_logits)
+        indices = exec_req.result_indices
 
         # Normal greedy selection based on max value
         if (top_k, top_p) == (None, None):
+            if indices is not None:
+                return indices.items[0]
```
Comment (rsuderman) [inline shortfin/python/shortfin_apps/llm/components/token_selection_strategy/greedy_token_selection_strategy.py:41]: You won't know this is true if `top_k` is set to `> 1` as topk returns the values unsorted. Indices should be proxied into `select_greedy` and used as n indirection. It should also be plumbed into the other token selection strategies.

File: sharktank/sharktank/examples/export_paged_llm_v1.py
Comment (rsuderman) [inline sharktank/sharktank/examples/export_paged_llm_v1.py:1]: You should also set the `top_k` value in the export json and check it `shortfin` side. This can be used to throw an error if the exported `top` selection falls outside of what we want to support service side.

**Review by rsuderman — CHANGES_REQUESTED** [review]

File: sharktank/sharktank/examples/export_paged_llm_v1.py [hunk 6]
```diff
@@ -305,6 +305,14 @@ def _(model, tokens, seq_lens, seq_block_ids, cs):
...
+                if args.top_k == 1:
+                    max_logits, indices = model.argmax(
+                        logits, chunk_size=hp.context_length // 128
+                    )
+
```
[the remark this answers is not in the document]
Comment (stbaione) [reply sharktank/sharktank/examples/export_paged_llm_v1.py:316]: Updated

File: sharktank/sharktank/examples/export_paged_llm_v1.py (hunk 1 above)
Comment (stbaione) [reply sharktank/sharktank/examples/export_paged_llm_v1.py]: Updated

File: sharktank/sharktank/examples/export_paged_llm_v1.py (hunk 2 above)
Comment (stbaione) [reply sharktank/sharktank/examples/export_paged_llm_v1.py]: Updated

File: sharktank/sharktank/ops/default_impls.py (hunk 3 above)
Comment (stbaione) [reply sharktank/sharktank/ops/default_impls.py:59]: Updated

File: shortfin/python/shortfin_apps/llm/components/token_selection_strategy/greedy_token_selection_strategy.py (hunk 5 above)
Comment (stbaione) [reply shortfin/python/shortfin_apps/llm/components/token_selection_strategy/greedy_token_selection_strategy.py:41]: For exporting the `topk` function, my thinking was I'll be setting the export to give us the sorted logits. Figured it'd be more efficient to get that over with on the GPU.

That would make it so that this line would always hold true

File: shortfin/python/shortfin_apps/llm/components/batcher.py (hunk 4 above)
Comment (stbaione) [reply shortfin/python/shortfin_apps/llm/components/batcher.py]: Updated

File: sharktank/sharktank/examples/export_paged_llm_v1.py
Comment (stbaione) [reply sharktank/sharktank/examples/export_paged_llm_v1.py:1]: Added error check for `top_k != 1`

File: shortfin/python/shortfin_apps/llm/components/token_selection_strategy/greedy_token_selection_strategy.py (hunk 5 above)
Comment (rsuderman) [reply shortfin/python/shortfin_apps/llm/components/token_selection_strategy/greedy_token_selection_strategy.py:41]: Lets not worry too much about making sure logits are sorted. CPU wise sorting ~100 objects is basically instant so its unlikely we will even both on the GPU.
